Add Standalone Production config

buti1021 opened the pull request:
Following the excellent work by Anthony I split the nginx into a proper config production which doesnt use the vite development server.

~~I still have a few open questions/remarks I want to discuss:~~
1. ~~Env File, imo we should put the .env into version control, just an example of it. My prefered solution would be to create a file called `.env-example` and .gitignore the `.env` file.~~ done, created example.env
2. ~~The development version as provided right now is not really working, as the vite can't access the wss for HMR. I would ask the question if we should really use nginx for the dev setup. Happy for your input here.~~ done, removed nginx for dev setup
3. ~~Healthchecks: @bendhouseart Your changes to the healthchecks in the compose file, didn't work for me. What was the point behind your changes?~~ done
4. ~~Startup scripts (dev.sh and prod.sh) I see the benefit of those files (git submodules & key generation) but also would try to look for a streamlining of the storage of environment variables (script vs. .env) and the duplication of code between the dev and prod file. + I don't completly understand why the npm dependencies are installed in that script and not in the Dockerfiles of the corresponding containers. Option might be to merge dev and prod into a start.sh with more flags.~~ done launch.sh is the entrypoint for now, configuration can be done via .env file
5. ~~Nginx config file, I think this a good starting point, not sure tho how far we can go with this as setups are quiet different (e.g. we have dont have our cert key encrypted and as a result don't have a ssl_password_file. Thus the nginx fails. Maybe remove the config from the code repo and put it into the documentation as an example instead?~~ done, sticking to  versioning the nginx config for now :)
6. ~~The api still runs in dev mode more or less. Didn't completly get what the start.sh is used for here (maybe nowhere at all?). Might have time to tackle that in the next few days.~~ this is still undone, but for now out of the scope of this PR.

Happy for your input.

Best,

Tim

bendhouseart committed:
updating docker compose to use nginx for forwarding

bendhouseart committed:
added nginx config and updated docker file works on mac

bendhouseart committed:
added folder for certs

bendhouseart committed:
had to bodge npm installs, husky, and vite to get healthy containers on osx

bendhouseart committed:
Merge branch 'add-nginx-forwarding' of github.com:bendhouseart/ezbids into add-nginx-forwarding

bendhouseart committed:
updated docker compose, health checks all pass. CORS still an issue

bendhouseart committed:
add nginx config

bendhouseart committed:
incremental progress is progress

bendhouseart committed:
redirects all work when running locally

bendhouseart committed:
Merge branch 'add-nginx-forwarding' of github.com:bendhouseart/ezbids into add-nginx-forwarding

bendhouseart committed:
ssl works with the addition of a cert, key, and the password for it

Commit:
create standalone production config

bendhouseart commented:
> Env File, imo we should put the .env into version control, just an example of it. My prefered solution would be to create a file called .env-example and .gitignore the .env file.

Yeah, that makes sense, but I prefer `example.env` ;), but I think it's worth adding an `if [ ! -f .env ]; then cp example.env .env fi` somewhere to handle things (adding this to my branch right now).

> The development version as provided right now is not really working, as the vite can't access the wss for HMR. I would ask the question if we should really use nginx for the dev setup. Happy for your input here.

I'd like to commit to nginx always being run b/c that makes the development and production environments as similar as possible (barring how brainlife.io deploys things). That is to say I think it's simpler to have 2 nginx configuration files and 1 env file then to have multiple docker compose files, but that solution while being simpler to me may not be the best path forward.

> Healthchecks: @bendhouseart Your changes to the healthchecks in the compose file, didn't work for me. What was the point behind your changes?

It was required as the health checks were failing when reaching out to `localhost` after we got `nginx` up and behaving. Since we were all in on nginx it seemed like a good idea to just go with it in the compose too.

> Startup scripts (dev.sh and prod.sh) I see the benefit of those files (git submodules & key generation) but also would try to look for a streamlining of the storage of environment variables (script vs. .env) and the duplication of code between the dev and prod file. + I don't completly understand why the npm dependencies are installed in that script and not in the Dockerfiles of the corresponding containers. Option might be to merge dev and prod into a start.sh with more flags.

Yeah, I had some similar remarks about the run/build process, but I think it's helpful to keep in mind that this is _not_ how ezbids is deployed on brainlife.io. Technically speaking, everything we've been doing has been "non-production" compared to how this project was originally scoped/intended to work. 

I agree and will be working today to streamline the launch/build process to better leverage:

- environment variables via .env and otherwise
- making `dev.sh` a bit more explicit by renaming it `launch.sh`
- adding more options to `dev.sh`/`launch.sh`  for production, rebuilding/installing npm after mounting etc.

> Nginx config file, I think this a good starting point, not sure tho how far we can go with this as setups are quiet different (e.g. we have dont have our cert key encrypted and as a result don't have a ssl_password_file. Thus the nginx fails. Maybe remove the config from the code repo and put it into the documentation as an example instead?

Yeah when I requested my ssl certs I encrypted the keys, I'm not sure of a clever way around this within the nginx config. Before we think too much on this, could you try running the `http` version (point to `nginx/development_nginx.conf`) instead? My thinking is to use the http nginx enabled version for local development and the https nginx version for production going forward. 

> The api still runs in dev mode more or less. Didn't completly get what the start.sh is used for here (maybe nowhere at all?). Might have time to tackle that in the next few days.

No idea either, but I think consolidation is the word of the week for my branch.

bendhouseart commented on ui/Dockerfile-production:
```diff
@@ -0,0 +1,9 @@
+
+RUN npm install
+
+CMD [ "npm", "run", "build"] 
```
So when I switch over to using the `build` option it seems to break my application and make it unusable. Do the options `build` and `dev` that get used in this docker file correspond to this -> https://github.com/brainlife/ezbids/blob/8ddb0a0291ac5a9d36b3ac175602245c35843383/ui/package.json#L4

Really not sure what exactly is all going on here.

bendhouseart replied:
What further confuses me is what effect this vite config has, see -> https://github.com/brainlife/ezbids/blob/8ddb0a0291ac5a9d36b3ac175602245c35843383/ui/vite.config.ts#L8-L20

Is this at all related to the page refreshing?

bendhouseart committed:
consolidating variables and options into 1 launch script and 1 .env file

bendhouseart committed:
updated lint check to use prettier instead of npm 'style-check' and 'lint-check'

bendhouseart committed:
update package-lock.json

bendhouseart committed:
bringing more in line with buti1021's PR, while 'stream-lining' env and entrypoints

bendhouseart committed:
update docker ignore and build

Commit:
works well enough in production, 'dev' could still use some tweaks

bendhouseart committed:
what changes?

bendhouseart committed:
Apply suggestions from code review

update nginx.conf

bendhouseart committed:
got local running again by ditching nginx

buti1021 committed:
remove nginx_var

buti1021 committed:
Update example.env

buti1021 committed:
remove last dev nginx bits

buti1021 committed:
certs and keys path as environment

buti1021 committed:
Merge branch 'add-nginx-forwarding' into add-nginx-forwarding

buti1021 committed:
enable telemetry via build in compose_profile env var

buti1021 committed:
revert vite config changes

buti1021 committed:
Merge pull request #1 from bendhouseart/add-nginx-forwarding

Add nginx forwarding

buti1021 committed:
remove old files

buti1021 commented:
Hey everyone, I've been working with @bendhouseart to improve the setup and deployment on production especially for a setup where client and handler run on different machines. I the context of that we:
1. Renamed dev.sh to launch.sh and moved all the configuration done via command line arguments to an .env file.
2. Created a docker-compose with nginx for production to allow easy setup with a default configuration.

We tested both setup on our local infrastructure. Please let me know if anything else is missing ot get this merged. (Not sure why the dev.sh file is marked as a conflict, this PR should remove the dev.sh file)

Best

bendhouseart commented:
Yeah, I went ahead an closed the original PR as they've converged now with this. Currently able to launch this in either "production" or "dev" mode and confident enough to say lgtm.
